library/core/src/marker.rs

```diff
@@ -879,25 +879,35 @@ marker_impls! {
         {T: ?Sized} &mut T,
 }
 
-/// Types that can be safely moved after being pinned.
+/// Types that do not need to follow the rules of pinning.
 ///
-/// Rust itself has no notion of immovable types, and considers moves (e.g.,
-/// through assignment or [`mem::replace`]) to always be safe.
+/// For information on what "pinning" is, see the [`pin` module] documentation.
 ///
-/// The [`Pin`][Pin] type is used instead to prevent moves through the type
-/// system. Pointers `P<T>` wrapped in the [`Pin<P<T>>`][Pin] wrapper can't be
-/// moved out of. See the [`pin` module] documentation for more information on
-/// pinning.
+/// Implementing the `Unpin` trait for `T` lifts the restrictions of pinning off that type.
+/// This means that, if `T: Unpin`, it cannot be assumed that a value of type `T` will be bound
+/// by the invariants that pinning infers, *even* when "pinned" by a [`Pin<Ptr>`] pointing at it.
+/// When a value of type `T` is pointed at by a [`Pin<Ptr>`], [`Pin`] will not restrict access
+/// to the pointee value like it normally would, thus allowing the user to do anything that they
+/// normally could with a non-[`Pin`]-wrapped `Ptr` to that value.
```

Comment:
I think this entire paragraph is backwards. Implementing this trait indicates that the type has a certain property. We certainly don't want people to think "oh I don't like those pinning restrictions, let me quickly implement that trait to silence the compiler". When considering whether you can make your type Unpin, you shouldn't consider Pin and its access restrictions you should consider: does my type have a separate pinning invariant, or is it completely pinning-agnostic? The fact that Pin then imposes no restriction is a *consequence*, it should not be the first point is the line of reasoning related to Unpin, but the last point.

Reply:
True. How would you suggest we amend the phrasing here to correct it? Maybe something along the following lines?

```diff
- Types that do not need to follow the rules of pinning.
+ Types that do not feature any pinning guarantees.
  
- Implementing the `Unpin` trait for `T` lifts the restrictions of pinning off that type.
- This means that, if `T: Unpin`, it cannot be assumed that a value of type `T` will be bound
- by the invariants that pinning infers, *even* when "pinned" by a [`Pin<Ptr>`] pointing at it.
+ Implementing the `Unpin` trait for `T` expresses the fact that `T` is pinning-agnostic:
+ it shall not expose nor rely on any pinning guarantees. This, in turn, means that a
+ `Pin`-wrapped pointer to such a type can feature a *fully unrestricted* API.
+ In other words, if `T: Unpin`, a value of type `T` will *not* be bound by the invariants
+ which pinning otherwise offers, even when "pinned" by a [`Pin<Ptr>`] pointing at it.
  When a value of type `T` is pointed at by a [`Pin<Ptr>`], [`Pin`] will not restrict access
  to the pointee value like it normally would, thus allowing the user to do anything that they
  normally could with a non-[`Pin`]-wrapped `Ptr` to that value.
```

Reply:
Nice, I see what you mean. Will try a new version of this along the lines of what @danielhenrymantilla wrote.  One tricky thing and the reason that I didn't do it this way originally is that I wanted to keep the introductory `Unpin` docs understandable to a user who does not fully understand pinning. Thus it's hard to use terms like "address-sensitive values" because a user may not understand that. I want to try to strike a balance where people who are just *consuming* APIs that involve `Pin` (for example, futures-related apis) without *writing them themselves* or *creating their own types* that interact with them can get an overview of what's happening without needing to deep dive and grok the whole thing. I do sympathise that we don't want to let that compromise the docs' value for the cases where someone actually *is* making their own type and thus needs to fully grok things to not make unsound mistakes.

Reply:
I think we absolutely should use terms like "address-sensitive values" here, it's important to relate this to the `pin` docs. Anyone *implementing* this trait should understand what an address-sensitive value is, or else they are likely to cause problems.

Are you worried about people that come here because they wondering about the trait bound and where to get that one from? I can see how maybe we could try to do something for those people that does not involve the full module-level docs. But we should make it very clear that before implementing the trait, an understanding of the entire concept is required! This will be surprising since it's a safe trait, so we should emphasize this heavily.

Reply:
I've applied the suggestions so far in https://github.com/rust-lang/rust/pull/116129/commits/1e6b21ed219479d86d435d51df22cba797d0159b

I think address-sensitiveness is a good thing to delve into that I would prefer we do as a followup.

Reply:
Works for me.